File: frontend/replace_scrolls.js
const fs = require('fs');
const path = require('path');

function getRelativePath(fromFile, targetPath) {
  let rel = path.relative(path.dirname(fromFile), targetPath);
  if (!rel.startsWith('.')) rel = './' + rel;
  rel = rel.replace(/\.tsx?$/, '');
  return rel;
}

function processDir(dir) {
  const files = fs.readdirSync(dir);
  for (const file of files) {
    const fullPath = path.join(dir, file);
    if (fs.statSync(fullPath).isDirectory()) {
      processDir(fullPath);
    } else if (fullPath.endsWith('.tsx') || fullPath.endsWith('.ts')) {
      processFile(fullPath);
    }
  }
}

function processFile(filePath) {
  let content = fs.readFileSync(filePath, 'utf8');
  
  if (!content.includes('<ScrollView') && !content.includes('<FlatList') && !content.includes('</ScrollView>') && !content.includes('</FlatList>')) return;
  if (content.includes('VerticalScrollView') || content.includes('VerticalFlatList')) return;

  const originalContent = content;

  // Let's replace </ScrollView> and </FlatList> globally
  content = content.replace(/<\/ScrollView>/g, '</VerticalScrollView>');
  content = content.replace(/<\/FlatList>/g, '</VerticalFlatList>');

  // Let's replace <ScrollView ...> and <FlatList ...>
  // BUT revert ones that have horizontal
  content = content.replace(/<ScrollView/g, '<VerticalScrollView');
  content = content.replace(/<FlatList/g, '<VerticalFlatList');

  // Fix the ones that actually have horizontal to use the original
  // E.g. <VerticalScrollView horizontal ...
  // This is a simpler logic: just change them back!
  content = content.replace(/<VerticalScrollView([^>]*)\bhorizontal\b/g, '<ScrollView$1horizontal');
  content = content.replace(/<VerticalFlatList([^>]*)\bhorizontal\b/g, '<FlatList$1horizontal');

  // Fix closing tags for horizontal ones! Wait, balancing tags using regex is hard.
  // We can just rely on the fact that horizontal scrollviews are usually self-closing or very small.
  // Let's check if there are any horizontal ScrollViews in the codebase.
  
  // To be safe, let's just use the AST / RegExp.
  
  if (content !== originalContent) {
    const hasVertScroll = content.includes('<VerticalScrollView') || content.includes('</VerticalScrollView>');
    const hasVertFlat = content.includes('<VerticalFlatList') || content.includes('</VerticalFlatList>');
    
    if (hasVertScroll || hasVertFlat) {
      const absWrapperFile = '/Users/hariharan/ritgate/frontend/src/components/navigation/VerticalScrollViews';
      const relPath = getRelativePath(filePath, absWrapperFile);
      
      let imports = [];
      if (hasVertScroll) imports.push('VerticalScrollView');
      if (hasVertFlat) imports.push('VerticalFlatList');
      
      const importStr = `import { ${imports.join(', ')} } from '${relPath}';\n`;
      
      const lines = content.split('\n');
      let lastImportIndex = -1;
      for (let i = 0; i < lines.length; i++) {
        if (lines[i].trim().startsWith('import ')) {
          lastImportIndex = i;
        }
      }
      
      if (lastImportIndex !== -1) {
        lines.splice(lastImportIndex + 1, 0, importStr);
        content = lines.join('\n');
      } else {
        content = importStr + content;
      }
      
      fs.writeFileSync(filePath, content, 'utf8');
      console.log('Processed', filePath);
    }
  }
}

processDir(__dirname + '/src/screens');

File: frontend/replace_scrolls.py
import os
import re

def process_file(filepath):
    with open(filepath, 'r') as f:
        content = f.read()

    if '<ScrollView' not in content and '<FlatList' not in content:
        return

    # Skip files that contain horizontal scrollviews
    if re.search(r'<ScrollView[^>]*\bhorizontal', content) or re.search(r'<FlatList[^>]*\bhorizontal', content):
        print(f"SKIPPING horizontal-mixed file: {filepath}")
        return

    # We must only replace actual tags, not imports. 
    # Because <ScrollView and </ScrollView> ARE tags, but what if there's no space?
    
    # We shouldn't replace the react-native import for ScrollView because they might use ScrollViewProps!
    # Instead, we just replace instances of `<ScrollView` with `<VerticalScrollView`
    original = content
    content = content.replace('<ScrollView', '<VerticalScrollView')
    content = content.replace('</ScrollView>', '</VerticalScrollView>')
    content = content.replace('<FlatList', '<VerticalFlatList')
    content = content.replace('</FlatList>', '</VerticalFlatList>')

    if content != original:
        rel_path = os.path.relpath('/Users/hariharan/ritgate/frontend/src/components/navigation/VerticalScrollViews', os.path.dirname(filepath))
        if not rel_path.startswith('.'):
            rel_path = './' + rel_path
        
        imports = []
        if 'VerticalScrollView' in content: imports.append('VerticalScrollView')
        if 'VerticalFlatList' in content: imports.append('VerticalFlatList')
        
        import_str = f"import {{ {', '.join(imports)} }} from '{rel_path}';\n"
        
        lines = content.split('\n')
        last_import = -1
        for i, line in enumerate(lines):
            if line.startswith('import '):
                last_import = i
                
        if last_import != -1:
            lines.insert(last_import + 1, import_str)
            content = '\n'.join(lines)
        else:
            content = import_str + content
            
        with open(filepath, 'w') as f:
            f.write(content)
        print(f"Processed: {filepath}")
    else:
        print(f"No changes made to {filepath}")

def process_dir(directory):
    for root, _, files in os.walk(directory):
        for file in files:
            if file.endswith('.tsx') or file.endswith('.ts'):
                process_file(os.path.join(root, file))

process_dir('/Users/hariharan/ritgate/frontend/src/screens')
